ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/core/tests/internal/XmlParserFactoryTest.java

```diff
@@ -0,0 +1,203 @@
+	}
+
+	@Test
+	public void testParseXmlWithIgnoredExternalEntity() throws Exception {
+		SAXParser parser = XmlParserFactory.createSAXParserIgnoringDOCTYPE();
+		testParseXmlWithExternalEntity(parser, this::createMalciousXml);
+	}
+
+	@Test
+	public void testParseXmlWithoutIgnoredExternalEntity() throws Exception {
+		SAXParser parser = XmlParserFactory.createSAXParserIgnoringDOCTYPE();
+		testParseXmlWithExternalEntity(parser, this::createNormalXml);
+	}
+
+	InputStream createMalciousXml(int localPort) {
+		try {
+			Path tempSecret = Files.createTempFile("test", ".txt");
+			tmpFiles.add(tempSecret);
+			Files.writeString(tempSecret, "secret");
+			Path tempDtd = Files.createTempFile("test", ".dtd");
+			tmpFiles.add(tempDtd);
+			String dtdContent = "<!ENTITY % var1 SYSTEM \"" + tempSecret.toUri().toURL() + "\">\n" //
+					+ "<!ENTITY var4 SYSTEM \"" + tempSecret.toUri().toURL() + "\">\n" //
+					+ "<!ENTITY % var2 \"<!ENTITY var3 SYSTEM 'http://localhost:" + localPort + "/?%var1;'>\">\n" //
+					+ "%var2;\n";
+			Files.writeString(tempDtd, dtdContent);
+			StringBuilder sb = new StringBuilder();
+			sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
+			URI dtdURi = tempDtd.toUri();
+			String dtdUrl = dtdURi.toURL().toString();
+			sb.append("<!DOCTYPE var1 SYSTEM \"" + dtdUrl + "\">\n");
+			sb.append("<Body>&var3;&var4;</Body>");
+			String xml = sb.toString();
+			return new ByteArrayInputStream(xml.getBytes(StandardCharsets.UTF_8));
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	InputStream createNormalXml(int localPort) {
+		StringBuilder sb = new StringBuilder();
+		sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
+		sb.append("<Body>hello</Body>");
+		String xml = sb.toString();
+		return new ByteArrayInputStream(xml.getBytes(StandardCharsets.UTF_8));
+	}
+
+	private void cleanup() throws IOException {
+		for (Path path : tmpFiles) {
+			Files.delete(path);
+		}
+	}
+
+	public void testParseXmlWithExternalEntity(SAXParser parser, IntFunction<InputStream> xmlSupplier)
+			throws Exception {
+		Collection<Throwable> exceptionsInOtherThreads = new ConcurrentLinkedQueue<>();
+		Thread httpServerThread = null;
+		try (ServerSocket serverSocket = new ServerSocket(0)) {
+			int localPort = serverSocket.getLocalPort();
+			httpServerThread = new Thread("httpServerThread") {
+				@Override
+				public void run() {
+					try (Socket socket = serverSocket.accept()) {
+						try (BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
+							String firstLine = in.readLine();
+							try (OutputStream outputStream = socket.getOutputStream()) {
+								outputStream.write("HTTP/1.1 200 OK\r\n".getBytes(StandardCharsets.UTF_8));
+							}
+							assertTrue(firstLine, firstLine.startsWith("GET"));
+							assertFalse("Server received secret: " + firstLine, firstLine.contains("secret")); // var3
+							assertFalse("Server was contacted", true);
```

Comment:
Looks like this is always throwing an exception.
Can you tell why the test passes nevertheless?

Reply:
One test expects an exception (but does not come so far because of another exception) while another test makes sure the code is not called at all.